File: tests/day09.rs
mod common;

use std::collections::HashSet;

fn read() -> CaveMap {
    let raw_heights: Vec<Vec<usize>> = common::read_test_input("data/day-09/input.txt")
        .iter()
        .map(|line| -> Vec<usize> { line.split("").flat_map(|c| c.parse().ok()).collect() })
        .clone()
        .collect();

    assert_eq!(100, raw_heights.len(), "expect n lines of input");

    for (index, row) in raw_heights.iter().enumerate() {
        assert_eq!(100, row.len(), "expect n cols in row #{}", index);
    }

    let mut rows: [[usize; CAVE_SIZE]; CAVE_SIZE] = [[0; CAVE_SIZE]; CAVE_SIZE];
    for (ri, row) in raw_heights.iter().enumerate() {
        for (ci, &col) in row.iter().enumerate() {
            rows[ri][ci] = col;
        }
    }

    return CaveMap { rows: rows };
}

const CAVE_SIZE: usize = 100;

type Point = (usize, usize);

struct CaveMap {
    rows: [[usize; CAVE_SIZE]; CAVE_SIZE],
}

impl CaveMap {
    pub fn height(&self, point: Point) -> usize {
        let (row, col) = point;
        return self.rows[row][col];
    }

    pub fn risk_level(&self, point: Point) -> usize {
        return self.height(point) + 1;
    }

    pub fn adjacents(&self, point: Point) -> [Option<Point>; 4] {
        let mut adjs: [Option<Point>; 4] = [None; 4]; // T R B L
        let (row, col) = point;
        if row <= CAVE_SIZE && col <= CAVE_SIZE {
            if row > 0 {
                adjs[0] = Some((row - 1, col));
            }
            if col < CAVE_SIZE - 1 {
                adjs[1] = Some((row, col + 1));
            }
            if row < CAVE_SIZE - 1 {
                adjs[2] = Some((row + 1, col));
            }
            if col > 0 {
                adjs[3] = Some((row, col - 1));
            }
        }
        adjs
    }

    pub fn all_points(&self) -> Vec<Point> {
        let mut all: Vec<Point> = Vec::new();
        for row in 0..CAVE_SIZE {
            for col in 0..CAVE_SIZE {
                all.push((row, col));
            }
        }
        all
    }

    pub fn low_points(&self) -> Vec<Point> {
        return self
            .all_points()
            .iter()
            .filter(|&c| {
                let height = self.height(*c);
                let lower: Vec<Point> = self
                    .adjacents(*c)
                    .iter()
                    .cloned()
                    .filter(|&a| a.is_some())
                    .filter(|&a| {
                        if let Some(adj) = a {
                            return self.height(adj) <= height;
                        } else {
                            return false;
                        }
                    })
                    .flat_map(|a| a)
                    .collect();
                return lower.is_empty();
            })
            .cloned()
            .collect();
    }

    fn _expand_basin(&self, basin_points: &mut HashSet<Point>, points: Vec<Point>) -> Vec<Point> {
        let to_insert: Vec<Point> = points
            .iter()
            .filter(|&p| self.height(*p) < 9 && !basin_points.contains(p))
            .cloned()
            .collect();
        for &point in to_insert.iter() {
            basin_points.insert(point);
        }
        return to_insert
            .iter()
            .flat_map(|&p| -> Vec<Point> {
                self.adjacents(p)
                    .iter()
                    .flat_map(|&a| {
                        if let Some(adj) = a {
                            return vec![adj];
                        } else {
                            return Vec::new();
                        }
                    })
                    .collect()
            })
            .collect();
    }
    pub fn expand_basin(&self, low_point: Point) -> Vec<Point> {
        let mut basin_points: HashSet<Point> = HashSet::new();
        let initial_input: Vec<Point> = vec![low_point];
        let mut expansion: Vec<Point> = self._expand_basin(&mut basin_points, initial_input);
        let mut expanded_basin = expansion.len() > 0;
        while expanded_basin {
            expansion = self._expand_basin(&mut basin_points, expansion);
            expanded_basin = expansion.len() > 0;
        }
        return basin_points.iter().cloned().collect();
    }
}

#[test]
fn day09part1() {
    let input: CaveMap = read();

    let low_points = input.low_points();
    assert_eq!(238, low_points.len(), "expect number of low points");

    let sum_risk_levels: usize = low_points
        .iter()
        .map(|&p| input.risk_level(p))
        .fold(0, |a, v| a + v);
    assert_eq!(570, sum_risk_levels, "expect sum of risk levels");
}

#[test]
fn day09part2() {
    let input: CaveMap = read();

    let low_points = input.low_points();
    assert_eq!(238, low_points.len(), "expect number of low points");

    let mut basin_sizes: Vec<usize> = low_points
        .iter()
        .map(|&p| input.expand_basin(p).len())
        .collect();

    basin_sizes.sort();
    let solution: usize = basin_sizes.iter().rev().take(3).fold(1, |a, v| a * v);

    assert_eq!(
        899392, solution,
        "expect solution for triple basin size product"
    );
}

File: ts-src/day09.ts
import { map, Observable, reduce } from "rxjs";
import { readTestInputLines } from "./util";

export const CAVE_SIZE = 100;
export type Point = { row: number, col: number };

export class PointSet {
    rows: boolean[][];
    constructor() {
        var rows: boolean[][] = [];
        for (var irow = 0; irow < CAVE_SIZE; irow++) {
            rows.push(Array.from({ length: CAVE_SIZE }, () => false));
        }
        this.rows = rows;
    }

    add(point: Point) {
        var { row, col } = point;
        this.rows[row][col] = true;
    }

    has(point: Point): boolean {
        var { row, col } = point;
        return this.rows[row][col];
    }

    values(caveMap: CaveMap): Point[] {
        return caveMap.allPoints().filter(p => this.has(p));
    }
}

export class CaveMap {
    rows: number[][];
    constructor(rows: number[][]) {
        this.rows = rows;
    }

    height(point: Point): number {
        var { row, col } = point;
        return this.rows[row][col];
    }

    riskLevel(point: Point): number {
        return this.height(point) + 1;
    }

    adjacents(point: Point): Point[] {
        var adjs: Point[] = [];
        var { row, col } = point;
        if (row <= CAVE_SIZE && col <= CAVE_SIZE) {
            if (row > 0) {
                adjs.push({ row: row - 1, col: col });
            }
            if (col < CAVE_SIZE - 1) {
                adjs.push({ row: row, col: col + 1 });
            }
            if (row < CAVE_SIZE - 1) {
                adjs.push({ row: row + 1, col: col });
            }
            if (col > 0) {
                adjs.push({ row: row, col: col - 1 });
            }
        }
        return adjs;
    }

    allPoints(): Point[] {
        return Array.from({ length: CAVE_SIZE }, (vr, ir) => ir)
            .flatMap((row: number) => {
                return Array.from({ length: CAVE_SIZE }, (vc, ic) => ic)
                    .map((col: number) => { return { row: row, col: col }; });
            });
    }

    lowPoints(): Point[] {
        return this.allPoints().filter((c) => {
            var height = this.height(c);
            var lower = this.adjacents(c).filter((adj) => this.height(adj) <= height);
            return lower.length == 0;
        });
    }

    _expandBasin(basinPoints: PointSet, points: Point[]): Point[] {
        var toInsert = points.filter(p => this.height(p) < 9 && !basinPoints.has(p))
        for (var i = 0; i < toInsert.length; i++) {
            if (toInsert[i]) {
                basinPoints.add(toInsert[i]);
            }
        }
        return toInsert.flatMap(p => this.adjacents(p));
    }

    expandBasin(lowPoint: Point): Point[] {
        var basinPoints = new PointSet();
        var expansion = this._expandBasin(basinPoints, [lowPoint]);
        var expandedBasin = expansion.length > 0;
        while (expandedBasin) {
            expansion = this._expandBasin(basinPoints, expansion);
            expandedBasin = expansion.length > 0;
        }
        return basinPoints.values(this);
    }
}

export const readInput = (): Observable<CaveMap> => readTestInputLines('day-09').pipe(
    map(line => line.split('').map(c => parseInt(c))),
    reduce((acc: number[][], row: number[]) => { acc.push(row); return acc; }, []),
    map(rows => new CaveMap(rows)),
);
